### PR TITLE
Add coercion.ml and coercion.mli

### PR DESCRIPTION
**Pull request by lukemaurer**

### CODE REVIEW

File: middle_end/flambda/compilenv_deps/coercion.mli [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+(*   Copyright 2019 Jane Street Group LLC                                 *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
```
Comment (mshinwell) [inline middle_end/flambda/compilenv_deps/coercion.mli:32]: Please add a comment describing what this file does.  Also remove 4 from the excluded warnings list.

File: middle_end/flambda/compilenv_deps/coercion.ml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+(*                   Mark Shinwell, Jane Street Europe                    *)
+(*                                                                        *)
+(*   Copyright 2019 Jane Street Group LLC                                 *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
```
Comment (mshinwell) [inline middle_end/flambda/compilenv_deps/coercion.ml:14]: Let's add `open! Int_replace_polymorphic_compare`.  Also please copy the warning stanza from the .mli file (having removed number 4 from the exclusion list) and fix any resulting warnings.

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: middle_end/flambda/compilenv_deps/coercion.mli (hunk 1 above)
Comment (lukemaurer) [reply middle_end/flambda/compilenv_deps/coercion.mli:32]: Okay, done.

File: middle_end/flambda/compilenv_deps/coercion.ml (hunk 2 above)
Comment (lukemaurer) [reply middle_end/flambda/compilenv_deps/coercion.ml:14]: Done (there was indeed a warning).

**Review by mshinwell — APPROVED** [review]

**Pull request merged**